### PR TITLE
APM-311115 Added validation for url and api token permissions

### PR DESCRIPTION
**Pull request by mswiatkowska**
Added the same url validation as we have in azure log forwarder deployment script.
Added API token permissions validation (the same validation added in other solutions).

### CODE REVIEW

File: dynatrace-aws-logs.sh [hunk 1]
```diff
@@ -158,7 +178,16 @@ arguments:
     TENANT_ID="" # NOT USED IN THIS CASE
   fi
 
+  if [[ "$USE_EXISTING_ACTIVE_GATE" == "false" ]] && ! [[ "${TARGET_URL}" =~ $DYNATRACE_TARGET_URL_REGEX ]]; then
+      echo "Invalid value for parameter --target-url. Example of valid url for deployment with ActiveGate: https://environment-id.live.dynatrace.com"
+      exit 1
+  elif [[ "$USE_EXISTING_ACTIVE_GATE" == "true" ]] && ! [[ "${TARGET_URL}" =~ $ACTIVE_GATE_TARGET_URL_REGEX ]]; then
+      echo "Invalid value for parameter --target-url. Example of valid url for deployment without ActiveGate: https://environemnt-active-gate-url:9999/e/environment-id"
```
Comment (mmajcher) [inline dynatrace-aws-logs.sh:185]: typo environment

**Review by mmajcher — DISMISSED** [review]

File: dynatrace-aws-logs.sh [hunk 2]
```diff
@@ -158,7 +178,16 @@ arguments:
...
+  if [[ "$USE_EXISTING_ACTIVE_GATE" == "false" ]] && ! [[ "${TARGET_URL}" =~ $DYNATRACE_TARGET_URL_REGEX ]]; then
+      echo "Invalid value for parameter --target-url. Example of valid url for deployment with ActiveGate: https://environment-id.live.dynatrace.com"
+      exit 1
+  elif [[ "$USE_EXISTING_ACTIVE_GATE" == "true" ]] && ! [[ "${TARGET_URL}" =~ $ACTIVE_GATE_TARGET_URL_REGEX ]]; then
+      echo "Invalid value for parameter --target-url. Example of valid url for deployment without ActiveGate: https://environment-active-gate-url:9999/e/environment-id"
+      exit 1
+  fi
+
   print_params_deploy
+  check_api_token
```
Comment (dt-be) [inline dynatrace-aws-logs.sh:190]: I'd move the check above printing params - for me printing params is like saying checks are already ok

File: dynatrace-aws-logs.sh [hunk 3]
```diff
@@ -88,6 +91,23 @@ arguments:
...
+  check_api_token() {
+    if RESPONSE=$(curl -k -s -X POST -d "{\"token\":\"$TARGET_API_TOKEN\"}" "$TARGET_URL/api/v2/apiTokens/lookup" -w "<<HTTP_CODE>>%{http_code}" -H "accept: application/json; charset=utf-8" -H "Content-Type: application/json; charset=utf-8" -H "Authorization: Api-Token $TARGET_API_TOKEN"); then
+      CODE=$(sed -rn 's/.*<<HTTP_CODE>>(.*)$/\1/p' <<<"$RESPONSE")
+      RESPONSE=$(sed -r 's/(.*)<<HTTP_CODE>>.*$/\1/' <<<"$RESPONSE")
+      if [ "$CODE" -ge 300 ]; then
+        echo "Failed to check Dynatrace API token permissions - please verify provided values for parameters: --target-url and --target-api-token. $RESPONSE"
+        exit 1
+      fi
+      if ! grep -q "logs.ingest" <<<"$RESPONSE"; then
+        echo "Missing Ingest logs permission for the API token"
```
Comment (dt-be) [inline dynatrace-aws-logs.sh:103]: I think you have recently been discussing to info if permissions are v1 or v2. I'd have it here as well (v2)

File: dynatrace-aws-logs.sh (hunk 2 above)
Comment (mswiatkowska) [reply dynatrace-aws-logs.sh:190]: Firstly I wanted to add checking api token before printing but in Azure solution as we have two modes there and we don't print params in interactive mode I needed to do checking right before deploying so in aws I did it in the same way to be consistent.
Anyone has the same feeling it should be before printing?

File: dynatrace-aws-logs.sh [hunk 4]
```diff
@@ -158,7 +178,16 @@ arguments:
     TENANT_ID="" # NOT USED IN THIS CASE
   fi
 
+  if [[ "$USE_EXISTING_ACTIVE_GATE" == "false" ]] && ! [[ "${TARGET_URL}" =~ $DYNATRACE_TARGET_URL_REGEX ]]; then
+      echo "Invalid value for parameter --target-url. Example of valid url for deployment with ActiveGate: https://environment-id.live.dynatrace.com"
+      exit 1
+  elif [[ "$USE_EXISTING_ACTIVE_GATE" == "true" ]] && ! [[ "${TARGET_URL}" =~ $ACTIVE_GATE_TARGET_URL_REGEX ]]; then
+      echo "Invalid value for parameter --target-url. Example of valid url for deployment without ActiveGate: https://environment-active-gate-url:9999/e/environment-id"
```
Comment (dt-be) [inline dynatrace-aws-logs.sh:185]: I'd consider having this as in the doc, but this is minor

```Determine your API URL, which depends on where you want to set up generic log ingestion.
If you want to install a containerized ActiveGate, set it to your Dynatrace SaaS environment: https://<your_environment_ID>.live.dynatrace.com
If you want to use an existing ActiveGate, set it to your ActiveGate endpoint: https://<your_activegate_IP_or_hostname>:9999/e/<your_environment_ID> Note: To determine <your_environment_ID>, see environment ID.
```

**Review by dt-be — APPROVED** [review]

**Review by mmajcher — APPROVED** [review]

File: dynatrace-aws-logs.sh [hunk 5]
```diff
@@ -18,8 +18,8 @@ readonly LAMBDA_ZIP_NAME="dynatrace-aws-log-forwarder-lambda.zip"
 
 readonly DEFAULT_STACK_NAME="dynatrace-aws-logs"
 
-readonly DYNATRACE_TARGET_URL_REGEX="^https:\/\/[-a-zA-Z0-9@:%._+~=]{1,256}$"
-readonly ACTIVE_GATE_TARGET_URL_REGEX="^https:\/\/[-a-zA-Z0-9@:%._+~=]{1,256}\/e\/[a-z0-9-]{1,36}$"
+readonly DYNATRACE_TARGET_URL_REGEX="^https:\/\/[-a-zA-Z0-9@:%._+~=]{1,256}[\/]{0,1}$"
+readonly ACTIVE_GATE_TARGET_URL_REGEX="^https:\/\/[-a-zA-Z0-9@:%._+~=]{1,256}\/e\/[-a-z0-9]{1,36}[\/]{0,1}$"
```
Comment (dt-be) [inline dynatrace-aws-logs.sh:22]: `If you want to have the new ActiveGate created during deployment, set it to your Dynatrace SaaS environment: https://<your_environment_ID>.live.dynatrace.com`
DYNATRACE_TARGET_URL_REGEX="^https:\/\/**[-a-zA-Z0-9@:%._+~=]{1,256}**[\/]{0,1}$"
> `[-a-zA-Z0-9@:%._+~=]{1,256}`

`If you want to use an existing ActiveGate, set it to your ActiveGate endpoint: https://<your_activegate_IP_or_hostname>:9999/e/<your_environment_ID>`
ACTIVE_GATE_TARGET_URL_REGEX="^https:\/\/[-a-zA-Z0-9@:%._+~=]{1,256}\/e\/**[-a-z0-9]{1,36}**[\/]{0,1}$"
> `[-a-z0-9]{1,36}`

aren't both of those same env/tenantId section with different regex patterns? first probably also has full dnsname, so just asking

Comment (dt-be) [reply dynatrace-aws-logs.sh:22]: [\/]{0,1}
=
[\/]?

https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html
(not the most apt reference for bash regex, but I do think this is universal)

File: dynatrace-aws-logs.sh [hunk 6]
```diff
@@ -92,19 +92,20 @@ arguments:
...
-    if RESPONSE=$(curl -k -s -X POST -d "{\"token\":\"$TARGET_API_TOKEN\"}" "$TARGET_URL/api/v2/apiTokens/lookup" -w "<<HTTP_CODE>>%{http_code}" -H "accept: application/json; charset=utf-8" -H "Content-Type: application/json; charset=utf-8" -H "Authorization: Api-Token $TARGET_API_TOKEN"); then
+    URL=$(echo "$TARGET_URL" | sed 's:/*$::')
+    if RESPONSE=$(curl -k -s -X POST -d "{\"token\":\"$TARGET_API_TOKEN\"}" "$URL/api/v2/apiTokens/lookup" -w "<<HTTP_CODE>>%{http_code}" -H "accept: application/json; charset=utf-8" -H "Content-Type: application/json; charset=utf-8" -H "Authorization: Api-Token $TARGET_API_TOKEN"); then
       CODE=$(sed -rn 's/.*<<HTTP_CODE>>(.*)$/\1/p' <<<"$RESPONSE")
       RESPONSE=$(sed -r 's/(.*)<<HTTP_CODE>>.*$/\1/' <<<"$RESPONSE")
       if [ "$CODE" -ge 300 ]; then
-        echo "Failed to check Dynatrace API token permissions - please verify provided values for parameters: --target-url (${TARGET_URL}) and --target-api-token. $RESPONSE"
+        echo -e "\e[93mWARNING: \e[37mFailed to check Dynatrace API token permissions - please verify provided values for parameters: --target-url (${TARGET_URL}) and --target-api-token. $RESPONSE"
```
Comment (dt-be) [inline dynatrace-aws-logs.sh:100]: Just for reference:
-e allows escape codes
https://en.wikipedia.org/wiki/ANSI_escape_code
93 is yellow
37 is white

**Review by dt-be — APPROVED** [review]

**Pull request merged**